test/test_charm.py

```diff
@@ -477,155 +496,120 @@ def _get_action_test_meta(cls):
-                {'a': {(1, 2): 'c'}},
-                {'a': {None: 'c'}},
-                {'aBc': 'd'}):
-            charm.res = bad_res
-
-            with pytest.raises(ValueError):
-                charm.on.foo_bar_action.emit(id='1')
-
-    def _test_action_event_defer_fails(self, cmd_type: str):
-
-        class MyCharm(ops.CharmBase):
-
-            def __init__(self, *args: typing.Any):
-                super().__init__(*args)
-                framework.observe(self.on.start_action, self._on_start_action)
-
-            def _on_start_action(self, event: ops.ActionEvent):
-                event.defer()
-
-        fake_script(self, f"{cmd_type}-get", """echo '{"foo-name": "name", "silent": true}'""")
-        self.meta = self._get_action_test_meta()
-
-        os.environ[f'JUJU_{cmd_type.upper()}_NAME'] = 'start'
-        framework = self.create_framework()
-        charm = MyCharm(framework)
-
-        with pytest.raises(RuntimeError):
-            charm.on.start_action.emit(id='2')
 
-    def test_action_event_defer_fails(self):
-        self._test_action_event_defer_fails('action')
+def test_action_events(request: pytest.FixtureRequest, fake_script: FakeScript):
+
+    class MyCharm(ops.CharmBase):
+
+        def __init__(self, *args: typing.Any):
+            super().__init__(*args)
+            framework.observe(self.on.foo_bar_action, self._on_foo_bar_action)
+            framework.observe(self.on.start_action, self._on_start_action)
 
-    def test_containers(self):
-        meta = ops.CharmMeta.from_yaml("""
+        def _on_foo_bar_action(self, event: ops.ActionEvent):
+            self.seen_action_params = event.params
+            event.log('test-log')
+            event.set_results({'res': 'val with spaces', 'id': event.id})
+            event.fail('test-fail')
+
+        def _on_start_action(self, event: ops.ActionEvent):
+            pass
+
+    _setup_test_action(fake_script.write)
+    meta = _get_action_test_meta()
+    framework = create_framework(request, meta=meta)
+    charm = MyCharm(framework)
+
+    events: typing.List[str] = list(MyCharm.on.events())  # type: ignore
+    assert 'foo_bar_action' in events
+    assert 'start_action' in events
+
+    action_id = "1234"
+    charm.on.foo_bar_action.emit(id=action_id)
+    assert charm.seen_action_params == {"foo-name": "name", "silent": True}
+    assert fake_script.calls() == [
+        ['action-get', '--format=json'],
+        ['action-log', "test-log"],
+        ['action-set', "res=val with spaces", f"id={action_id}"],
+        ['action-fail', "test-fail"],
+    ]
+
+
+@pytest.mark.parametrize("bad_res", [
+    {'a': {'b': 'c'}, 'a.b': 'c'},
+    {'a': {'B': 'c'}},
+    {'a': {(1, 2): 'c'}},
+    {'a': {None: 'c'}},
+    {'aBc': 'd'}
+])
+def test_invalid_action_results(request: pytest.FixtureRequest,
+                                fake_script: FakeScript,
+                                bad_res: typing.Dict[str, typing.Any]):
```

Comment:
Shall we format all of these multiline ones as Ruff/Black does, to avoid further reformatting in future?

```suggestion
def test_invalid_action_results(
    request: pytest.FixtureRequest,
    fake_script: FakeScript,
    bad_res: typing.Dict[str, typing.Any],
):
```